pennylane/workflow/run.py

```diff
@@ -0,0 +1,317 @@
+
+    except ImportError as e:  # pragma: no cover
+        raise qml.QuantumFunctionError(
+            f"{interface} not found. Please install the latest "
+            f"version of {interface} to enable the '{interface}' interface."
+        ) from e
+
+    return ml_boundary
+
+
+def _make_inner_execute(device, inner_transform, execution_config=None) -> Callable:
+    """Construct the function responsible for executing quantum tapes within a ML framework boundary
+    for first-order derivatives.
+
+    The process involves performing device-specific preprocessing on the tapes (for new devices) or
+    device expansion (for legacy devices).
+
+    For higher-order derivatives, this function will delegate to another ML framework execution.
+    """
+
+    def inner_execute(tapes: "qml.tape.QuantumScriptBatch", **_) -> ResultBatch:
+        """Execution that occurs within a ML framework boundary.
+
+        Closure Variables:
+            expand_fn (Callable[[QuantumTape], QuantumTape]): A device preprocessing step
+            device (qml.devices.Device)
+        """
+
+        transformed_tapes, transform_post_processing = inner_transform(tapes)
+
+        if transformed_tapes:
+            results = device.execute(transformed_tapes, execution_config=execution_config)
+        else:
+            results = ()
+
+        return transform_post_processing(results)
+
+    return inner_execute
+
+
+# pylint: disable=too-many-branches
+def run(
+    tapes: "qml.tape.QuantumScriptBatch",
+    device: "qml.devices.Device",
+    resolved_execution_config: "qml.devices.ExecutionConfig",
+    inner_transform_program: TransformProgram,
+) -> ResultBatch:
+    """Execute a batch of quantum scripts on a device with optional gradient computation.
+
+    Args:
+        tapes (qml.tape.QuantumScriptBatch): batch of quantum scripts
+        device (qml.devices.Device): a Pennylane device
+        resolved_execution_config (qml.devices.ExecutionConfig): Configuration detailing
+            execution and differentiation settings.
+        inner_transform_program (TransformProgram): The transformation program to apply
+            to the quantum scripts before execution.
+
+    Returns:
+        ResultBatch: results of the execution
+    """
+    inner_execute = _make_inner_execute(device, inner_transform_program, resolved_execution_config)
+
+    # Exiting early if we do not need to deal with an interface boundary
+    no_interface_boundary_required = (
+        resolved_execution_config.interface == Interface.NUMPY
+        or resolved_execution_config.gradient_method
+        in {
+            None,
+            "backprop",
+        }
+    )
+    if no_interface_boundary_required:
+        results = inner_execute(tapes)
+        return results
+
+    if resolved_execution_config.interface != Interface.TF_AUTOGRAPH:
+        jpc, execute_fn = _construct_ml_execution_pipeline(
+            resolved_execution_config, device, inner_transform_program
+        )
```

Comment:
think we could move the other `TF_AUTOGRAPH` code here without breaking anything?

Reply:
Trying in latest commit 🤞🏼

Reply:
Seems to work! 🥳 Resolving.